File: devops-scm-api/src/main/java/com/tencent/devops/scm/api/pojo/webhook/git/PullRequestCommentHook.java
package com.tencent.devops.scm.api.pojo.webhook.git;

import com.tencent.devops.scm.api.pojo.PullRequest;
import java.util.Collections;
import java.util.Map;

import com.tencent.devops.scm.api.pojo.Review;
import lombok.AllArgsConstructor;
import lombok.Data;
import lombok.EqualsAndHashCode;
import lombok.NoArgsConstructor;
import lombok.experimental.SuperBuilder;

/**
 * pr或mr评论事件
 */
@Data
@AllArgsConstructor
@NoArgsConstructor
@SuperBuilder
@EqualsAndHashCode(callSuper = false)
public class PullRequestCommentHook extends AbstractCommentHook {
    public static final String CLASS_TYPE = "pull_request_comment";
    private PullRequest pullRequest;
    private Review review; // 工蜂commit review 专用

    @Override
    public Map<String, Object> outputs() {
        Map<String, Object> outputs = super.outputs();
        if (getExtras() != null) {
            outputs.putAll(getExtras());
        }
        // commit review
        if (review != null) {
            outputs.put("BK_REPO_GIT_WEBHOOK_REVIEW_STATE", review.getState().value);
            outputs.put("BK_REPO_GIT_WEBHOOK_REVIEW_ID", review.getId());
            outputs.put("BK_REPO_GIT_WEBHOOK_REVIEW_IID", review.getIid());
            outputs.put("BK_REPO_GIT_WEBHOOK_REVIEW_SOURCE_BRANCH", review.getSourceBranch());
            outputs.put("BK_REPO_GIT_WEBHOOK_REVIEW_SOURCE_PROJECT_ID", review.getSourceProjectId());
            outputs.put("BK_REPO_GIT_WEBHOOK_REVIEW_SOURCE_COMMIT", review.getSourceCommit());
            outputs.put("BK_REPO_GIT_WEBHOOK_REVIEW_TARGET_COMMIT", review.getTargetCommit());
            outputs.put("BK_REPO_GIT_WEBHOOK_REVIEW_TARGET_BRANCH", review.getTargetBranch());
            outputs.put("BK_REPO_GIT_WEBHOOK_REVIEW_TARGET_PROJECT_ID", review.getTargetProjectId());
        }
        return outputs;
    }
}

File: devops-scm-api/src/main/kotlin/com/tencent/devops/scm/api/pojo/webhook/git/PullRequestCommentHook.kt
package com.tencent.devops.scm.api.pojo.webhook.git

import com.tencent.devops.scm.api.constant.WebhookOutputCode.BK_REPO_GIT_WEBHOOK_REVIEW_ID
import com.tencent.devops.scm.api.constant.WebhookOutputCode.BK_REPO_GIT_WEBHOOK_REVIEW_IID
import com.tencent.devops.scm.api.constant.WebhookOutputCode.BK_REPO_GIT_WEBHOOK_REVIEW_SOURCE_BRANCH
import com.tencent.devops.scm.api.constant.WebhookOutputCode.BK_REPO_GIT_WEBHOOK_REVIEW_SOURCE_COMMIT
import com.tencent.devops.scm.api.constant.WebhookOutputCode.BK_REPO_GIT_WEBHOOK_REVIEW_SOURCE_PROJECT_ID
import com.tencent.devops.scm.api.constant.WebhookOutputCode.BK_REPO_GIT_WEBHOOK_REVIEW_STATE
import com.tencent.devops.scm.api.constant.WebhookOutputCode.BK_REPO_GIT_WEBHOOK_REVIEW_TARGET_BRANCH
import com.tencent.devops.scm.api.constant.WebhookOutputCode.BK_REPO_GIT_WEBHOOK_REVIEW_TARGET_COMMIT
import com.tencent.devops.scm.api.constant.WebhookOutputCode.BK_REPO_GIT_WEBHOOK_REVIEW_TARGET_PROJECT_ID
import com.tencent.devops.scm.api.enums.EventAction
import com.tencent.devops.scm.api.pojo.Comment
import com.tencent.devops.scm.api.pojo.PullRequest
import com.tencent.devops.scm.api.pojo.Review
import com.tencent.devops.scm.api.pojo.User
import com.tencent.devops.scm.api.pojo.repository.git.GitScmServerRepository

data class PullRequestCommentHook(
    override val action: EventAction,
    override val repo: GitScmServerRepository,
    override val eventType: String,
    override val comment: Comment,
    override val sender: User,
    override var extras: MutableMap<String, Any> = mutableMapOf(),
    val pullRequest: PullRequest? = null,
    val review: Review? = null
) : AbstractCommentHook(
    action = action,
    repo = repo,
    eventType = eventType,
    comment = comment,
    sender = sender,
    extras = extras
) {
    companion object {
        const val CLASS_TYPE = "issue_comment"
    }

    override val userName = sender.name

    override fun outputs(): Map<String, Any> {
        val outputs = super.outputs().toMutableMap()

        // 处理review信息
        review?.let {
            outputs[BK_REPO_GIT_WEBHOOK_REVIEW_STATE] = it.state?.value ?: ""
            outputs[BK_REPO_GIT_WEBHOOK_REVIEW_ID] = it.id
            outputs[BK_REPO_GIT_WEBHOOK_REVIEW_IID] = it.iid
            outputs[BK_REPO_GIT_WEBHOOK_REVIEW_SOURCE_BRANCH] = it.sourceBranch ?: ""
            outputs[BK_REPO_GIT_WEBHOOK_REVIEW_SOURCE_PROJECT_ID] = it.sourceProjectId ?: ""
            outputs[BK_REPO_GIT_WEBHOOK_REVIEW_SOURCE_COMMIT] = it.sourceCommit ?: ""
            outputs[BK_REPO_GIT_WEBHOOK_REVIEW_TARGET_COMMIT] = it.targetCommit ?: ""
            outputs[BK_REPO_GIT_WEBHOOK_REVIEW_TARGET_BRANCH] = it.targetBranch ?: ""
            outputs[BK_REPO_GIT_WEBHOOK_REVIEW_TARGET_PROJECT_ID] = it.targetProjectId ?: ""
        }

        if (extras.isNotEmpty()) {
            outputs.putAll(extras)
        }
        return outputs
    }
}
